Vectorized unbounded continuous random number generators (Issue 45)

bbbales2 opened the pull request:
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

This implements vectorized versions of all the unbounded continuous random number generators (with tests! Hooray!)

#### Details

I had to make some changes to test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp to allow for anywhere from one to three parameters to be tested.

I also had to update the way the Skew normal random numbers were getting generated to avoid infs (this is part of Issue #527)

I also fixed some indents in test/unit/math/prim/scal/prob/util.hpp and added an extra Google test check there (that will throw if the code tries to do something that leads to segfaults in the tests).

@seantalts Hope you were looking forward to checking some repetitive code

#### How to Verify:

```bash
./runTests.py test/unit/math/prim/mat/prob/normal_test.cpp
./runTests.py test/unit/math/prim/mat/prob/exp_mod_normal_test.cpp
./runTests.py test/unit/math/prim/mat/prob/skew_normal_test.cpp
./runTests.py test/unit/math/prim/mat/prob/student_t_test.cpp
./runTests.py test/unit/math/prim/mat/prob/cauchy_test.cpp
./runTests.py test/unit/math/prim/mat/prob/double_exponential_test.cpp
./runTests.py test/unit/math/prim/mat/prob/logistic_test.cpp
./runTests.py test/unit/math/prim/mat/prob/gumbel_test.cpp
```

And the scalar tests (though I didn't change these)
```bash
./runTests.py test/unit/math/prim/scal/prob/normal_test.cpp
./runTests.py test/unit/math/prim/scal/prob/exp_mod_normal_test.cpp
./runTests.py test/unit/math/prim/scal/prob/skew_normal_test.cpp
./runTests.py test/unit/math/prim/scal/prob/student_t_test.cpp
./runTests.py test/unit/math/prim/scal/prob/cauchy_test.cpp
./runTests.py test/unit/math/prim/scal/prob/double_exponential_test.cpp
./runTests.py test/unit/math/prim/scal/prob/logistic_test.cpp
./runTests.py test/unit/math/prim/scal/prob/gumbel_test.cpp
```

#### Side Effects:

Hopefully none

#### Documentation:

It's all inline

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): University of California, Santa Barbara

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

bbbales2 commented:
Poor ol' Travis. Package manager failed for some reason it seems.

bob-carpenter commented:
This is failing in unit testing because a new signed comparison warning cropped up in the test code.  Easy fix to change an `int` to a `size_t`.

bbbales2 commented:
Bah, I use clang++ for the nice errors, but I should probably just switch to g++ cause it seems to be more careful about these things. Tnx for the heads up.

bbbales2 commented:
I think this is ready to review. Jenkins failed on some upstream tests for some mystery reason:

```
Build step 'Execute shell' marked build as failure
ERROR: Step ‘Scan for compiler warnings’ failed: no workspace for Math Pull Request - Upstream - Stan #450
ERROR: Step ‘Publish JUnit test result report’ failed: no workspace for Math Pull Request - Upstream - Stan #450
ERROR: gelman-group-linux is offline; cannot locate JDK8u66
```

Could easily ask for a retest, but that'll burn another pile of hours on Jenkins for nothing. (no way this gets through without revisions anyway)

seantalts commented:
Sorry, the linux node has been tricksy so typically we'll just retest when this happens. I'm working on getting it to be more friendly. Bob, are you reviewing this or should someone else (me?)?

bbbales2 commented:
> tricksy

I just made it through the Hobbit again. Fantastic book. I'm totally into a Golem rebranding of Jenkins.

Maybe if we wanted a rebuild we'd haves to answers one of precious's riddles?

bbbales2 commented:
Jenkins retest this please

bbbales2 commented:
Jenkins retest this please

bbbales2 commented:
Nice! Tests passed with the gcc-4.9 patch in place. This is ready for review.

bob-carpenter commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -177,48 +253,151 @@ void check_dist_throws(T_generate_samples generate_samples,
+  check_dist_throws<RowVectorXd, VectorXd, RowVectorXd>
+    (generate_samples, good_p1, bad_p1, good_p2, bad_p2, good_p3, bad_p3);
+  check_dist_throws<RowVectorXd, RowVectorXd, RowVectorXd>
+    (generate_samples, good_p1, bad_p1, good_p2, bad_p2, good_p3, bad_p3);
```
Whenever you see this much redundant code, you want to ask yourself if it's really necessary.  

My first inclination here would be to wrap up the arguments `(generate_samples, good_p1, bad_p1, good_p2, bad_p2, good_p3, bad_p3)` into a structure.  Then it's just `check_dist_throws<T1, T2, T3>(s)` where `s` is the holding structure.  That also simplifies the argument declaration for `check_dist_throws`.   I think you should make this change.

I think you also want to be checking `int`---it can sometimes cause problems you don't see with `double` with argument ambiguity resolution.  I also think you should make this change.

My second inclination would be to use typelists to iterate over all the types.  It's a template programming technique worth learning and a little recursion will deal with all the permutations in a way that's probably easier to validate is correct.  The problem with these big lists is that it's very hard to check that they're complete.  And they're hard to maintain when you add a new type.  This change is optional, but I'd encourage you to at least check out the template technique and start looking over your shoulder every time you think about creating long lists like this.

bbbales2 commented:
Hopefully the tests pass. I changed stuff around quite a bit to address the comments.

> My first inclination here would be to wrap up the arguments (generate_samples, good_p1, bad_p1, good_p2, bad_p2, good_p3, bad_p3) into a structure. 

I changed how _rngs are tested. Now in each test you inherit from an abstract base class and build a special little object that has the right test hooks (see: https://github.com/bbbales2/math/blob/275889c5ba9185ea8cf6ec70b516013e4f63cb6f/test/unit/math/prim/mat/prob/normal_test.cpp). It's not actually a totally correct abstract base class cause the templated member function can't be virtual, but I thought this makes the test code look nice.

> My second inclination would be to use typelists to iterate over all the types.

Hopefully done. There were a bunch of options. I went with a recursive templating solution. There was some variadic templating stuff but it was beyond me. It was tricky to get the right combination of functors/templating so that all the pieces fit together, but it's not so bad now.

I did not add integer tests (or std::vector<int> tests). I guess it's important and I should do some thinking, but I might need to add extra lists of good/bad integer values. Rounding the good/bad double values won't always work (cause rounding a good value might give you an invalid value).

"vector_rng_test_helper.hpp" is starting to get pretty verbose with all its little helpers and functions in it. Should I be breaking this out into pieces? If so, any suggestions on what needs broken out?

I imagine this stuff is going to keep growing as the rest of the distribution things get added, so maybe it's worth waiting to see if any other big structural changes come along before we go through the process of breaking things out.

bob-carpenter commented:
> On Oct 1, 2017, at 6:29 PM, Ben Bales <notifications@github.com> wrote:
> 
> Hopefully the tests pass. I changed stuff around quite a bit to address the comments.
> 
> My first inclination here would be to wrap up the arguments (generate_samples, good_p1, bad_p1, good_p2, bad_p2, good_p3, bad_p3) into a structure.
> 
> I changed how _rngs are tested. Now in each test you inherit from an abstract base class and build a special little object that has the right test hooks (see: https://github.com/bbbales2/math/blob/275889c5ba9185ea8cf6ec70b516013e4f63cb6f/test/unit/math/prim/mat/prob/normal_test.cpp). It's not actually a totally correct abstract base class cause the templated member function can't be virtual, but I thought this makes the test code look nice.

I just brought up this same issue w.r.t. the "base class" we want for whole models.  Not
sure what the right answer is without being able to have virtual template methods.

> 
> My second inclination would be to use typelists to iterate over all the types.
> 
> Hopefully done. There were a bunch of options. I went with a recursive templating solution. There was some variadic templating stuff but it was beyond me. It was tricky to get the right combination of functors/templating so that all the pieces fit together, but it's not so bad now.

It's always a pain to do this.


> I did not add integer tests (or std::vector tests). I guess it's important and I should do some thinking, but I might need to add extra lists of good/bad integer values. Rounding the good/bad double values won't always work (cause rounding a good value might give you an invalid value).

Right---it needs an extra list of good/bad integer values.   And yes, this is important in that not having integers compile properly has bit us many times in the past.  We'll have to make sure that the tests in Stan itself at least instantiate the integer cases.  


> "vector_rng_test_helper.hpp" is starting to get pretty verbose with all its little helpers and functions in it. Should I be breaking this out into pieces? If so, any suggestions on what needs broken out?

Always break into pieces if you can.


> I imagine this stuff is going to keep growing as the rest of the distribution things get added, so maybe it's worth waiting to see if any other big structural changes come along before we go through the process of breaking things out.

Everything will keep growing, but that shouldn't freeze us in the meantime.

bob-carpenter commented:
For (1), Cherry pick results for now.  File an issue with Boost if
you have examples that fail.  (2) sounds good---we have a lot of
test harness stuff that could stand to be consolidated.

bbbales2 commented:
This is ready to be checked

syclik commented:
@bob-carpenter, I think @bbbales2 has made all the requested changes. Want to make sure and approve after you're done?

bob-carpenter commented:
Not everything got fixed.  I don't have time to finish the next review now---I'm only a couple of my comments in but I should be able to get to it over the next few days.

bbbales2 commented:
Cool beans, whenever is good. It'll be good to get this rolling again.

bob-carpenter commented:
I'll take another pass as soon as I can.  Sorry for the delay.

bob-carpenter commented on stan/math/prim/scal/prob/exp_mod_normal_rng.hpp:
```diff
@@ -3,38 +3,70 @@
-                       double sigma,
-                       double lambda,
+    /**
+     * Return a pseudorandom Exponentially modified normal variate for the
```
[optional fix]

I'd rather not capitalize anything other than proper names and names of distributions (not named after people) don't count.

bob-carpenter commented on stan/math/prim/scal/prob/exp_mod_normal_rng.hpp:
```diff
@@ -3,38 +3,70 @@
+     * given location, scale, and inverse scale using the specified random
+     * number generator.
+     *
+     * mu, sigma, and lambda can each be either scalars or vectors. All vector
```
Can't they be scalars, standard vectors, or Eigen vectors?

bob-carpenter commented on stan/math/prim/scal/prob/exp_mod_normal_rng.hpp:
```diff
@@ -3,38 +3,70 @@
+     * @return Exponentially modified normal random variate
+     * @throw std::domain_error if mu is infinite, sigma is nonpositive,
+     * or lambda is nonpositive
+     * @throw std::invalid_argument if vector arguments are not the same length
```
So I think this should say "container arguments" where you have "vector arguments".

bob-carpenter commented on stan/math/prim/scal/prob/exp_mod_normal_rng.hpp:
```diff
@@ -3,38 +3,70 @@
-        + exponential_rng(lambda, rng);
-    }
+      size_t N = max_size(mu, sigma, lambda);
+
```
[optional fix]

I'd remove all the vertical space that's not absolutely necessary for readability.  And that includes braces around for loops iwth single statements.  Vertical space is precious!

bob-carpenter commented on stan/math/prim/scal/prob/exp_mod_normal_rng.hpp:
```diff
@@ -3,38 +3,70 @@
 
+      for (size_t n = 0; n < N; n++) {
+        output[n] = normal_rng(mu_vec[n], sigma_vec[n], rng) +
+          exponential_rng(lambda_vec[n], rng);
```
[optional fix]

I promised Sean not to require changes not legislated by cpplint, but I really think opertors should go on the following line for readability.  You read code by scanning the structure down the left and it's lost with operators at theend of lines.

bob-carpenter commented on stan/math/prim/scal/prob/exp_mod_normal_rng.hpp:
```diff
@@ -3,38 +3,70 @@
+
+      VectorBuilder<true, double, T_loc, T_scale, T_inv_scale> output(N);
 
+      for (size_t n = 0; n < N; n++) {
```
[optional fix]

we prefer `++n` unless you need the result of `n++`  (again, this is in Google's style guide, but not enforced by cpplint)

bob-carpenter commented on stan/math/prim/scal/prob/gumbel_rng.hpp:
```diff
@@ -3,50 +3,64 @@
+    template <typename T_loc, typename T_scale, class RNG>
+    inline typename VectorBuilder<true, double, T_loc, T_scale>::type
+    gumbel_rng(const T_loc& mu,
+               const T_scale& beta,
```
[optional fix]

pack as many arguments onto a line as possible.  Remember, vertical space is precious.  (again, in Google style guide, not in cpplint)

bob-carpenter commented on stan/math/prim/scal/prob/gumbel_rng.hpp:
```diff
@@ -3,50 +3,64 @@
+
+      variate_generator<RNG&, uniform_01<> > uniform01_rng(rng, uniform_01<>());
+      for (size_t n = 0; n < N; n++) {
+        output[n] = mu_vec[n] -
```
[optional]

Same comments as last time on vertical space and operators.

bob-carpenter commented on stan/math/prim/scal/prob/logistic_rng.hpp:
```diff
@@ -3,26 +3,37 @@
+    template <typename T_loc, typename T_scale, class RNG>
+    inline typename VectorBuilder<true, double, T_loc, T_scale>::type
+    logistic_rng(const T_loc& mu,
+                 const T_scale& sigma,
```
[optional]

pack arguments up.  I'm going to stop repeating this, but it'd be great for me if it got fixed everywhere.

bob-carpenter commented on stan/math/prim/scal/prob/logistic_rng.hpp:
```diff
@@ -31,12 +42,25 @@ namespace stan {
       check_finite(function, "Location parameter", mu);
       check_positive_finite(function, "Scale parameter", sigma);
+      check_consistent_sizes(function,
+                             "Location parameter", mu,
```
[optional]

same comment here about packing arguments as much as possible

bob-carpenter commented on stan/math/prim/scal/prob/skew_normal_rng.hpp:
```diff
@@ -1,39 +1,79 @@
+      VectorBuilder<true, double, T_loc, T_scale, T_shape> output(N);
+
+      variate_generator<RNG&, normal_distribution<> >
+        norm_rng(rng, normal_distribution<>(0.0, 1.0));
```
[optional]

In general, we prefer `0` and `1` to `0.0` and `1.0` (yet another thing in Google style not caught by cpplint)

bob-carpenter commented on stan/math/prim/scal/prob/skew_normal_rng.hpp:
```diff
@@ -1,39 +1,79 @@
+        norm_rng(rng, normal_distribution<>(0.0, 1.0));
+      for (size_t n = 0; n < N; n++) {
+        double r1 = norm_rng(),
+          r2 = norm_rng();
```
[optional]

I'd prefer to have one declaration per line.  And of course, tighten up vertical spaces!

bob-carpenter commented on stan/math/prim/scal/prob/student_t_rng.hpp:
```diff
@@ -1,46 +1,72 @@
+     * Return a pseudorandom student-t variate for the given degrees of freedom,
+     * location, and scale using the specified random number generator.
+     *
+     * nu, mu, and sigma can each be either scalars or vectors. All vector
```
same fix here as everywhere about scalar/standard vector/Eigen vector.

bob-carpenter commented on test/unit/math/prim/mat/prob/VectorRNGTestRig.hpp:
```diff
@@ -0,0 +1,255 @@
+                   std::vector<double> good_p2,
+                   std::vector<int> good_p2_int,
+                   std::vector<double> bad_p2,
+                   std::vector<int> bad_p2_int) : N_(N), M_(M),
```
[optional]

Put the `:` on a new line.  You also don't want to waste horizontal space like this.

bbbales2 commented:
> Can't they be scalars, standard vectors, or Eigen vectors?
> So I think this should say "container arguments" where you have "vector arguments".

How about I change "vector" to "vector type" (like in https://github.com/stan-dev/stan/blob/develop/src/stan/lang/function_signatures.h#L10). This is how I did it in the normal_rng (must have just not copied it to the other _rngs yet):
```mu and sigma can each be either a scalar or vector type (the ones that scalar_seq_view knows about). Any vector inputs must be the same length.```

I'd like an easy way to refer to vector-like objects in a vague MATLAB/R this-is-vectorized kind of way.

> same comment here about packing arguments as much as possible

Sure thing, this should all be pretty easy. I'll go through and see how much I can compress things.

bob-carpenter commented:
How about "standard vector, Eigen vector, and Eigen row vector"?  That's explicit and not too long.  
The reason to avoid reference to other classes is to preserve encapsulation---those other classes may change without this one changing.




> On Oct 25, 2017, at 3:08 PM, Ben Bales <notifications@github.com> wrote:
> 
> Can't they be scalars, standard vectors, or Eigen vectors?
> So I think this should say "container arguments" where you have "vector arguments".
> 
> How about I change "vector" to "vector type" (like in https://github.com/stan-dev/stan/blob/develop/src/stan/lang/function_signatures.h#L10). This is how I did it in the normal_rng (must have just not copied it to the other _rngs yet):
> mu and sigma can each be either a scalar or vector type (the ones that scalar_seq_view knows about). Any vector inputs must be the same length.
> 
> I'd like an easy way to refer to vector-like objects in a vague MATLAB/R this-is-vectorized kind of way.
> 
> same comment here about packing arguments as much as possible
> 
> Sure thing, this should all be pretty easy. I'll go through and see how much I can compress things.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bbbales2 commented:
I gotta ask for help on this one. I'm not sure what's going on, but it feels fishy. @seantalts or @syclik -- you guys might recognize something here -- it doesn't really have much to do with the actual pull.

The first problem I had was the scalar tests for double_exponential_rng were failing when the vector ones weren't.

I found two ways to fix it (and make the scalar double_exponential tests pass).

1. Replace log with std::log (https://github.com/stan-dev/math/pull/622/commits/9bf8a0ea2d0dd227c1d2c2e66cce728d32cce725).

The effect I was seeing without this is that log(abs(z)) was going to -Inf, regardless of what z was. It felt like double was being cast to int, but I don't know if that should happen.

My guess was that it was only seeing the log definition here: https://github.com/stan-dev/math/blob/develop/stan/math/prim/scal/fun/log.hpp ? Which is only for ints? And for some reason wasn't seeing the usual std::log definition? But this doesn't explain why the vector tests worked.

2. Replace the stan/math/prim/scal.hpp include with stan/math/prim/mat.hpp

Both of these feel like wrong sorts of fixes. Any idea what could be going on?

seantalts commented:
I took a 5 minute look and have only one possible suggestion - the old code had a `using std::log;` statement - was that present still when you saw the -Inf? It seems from my brief reading of it like that `using` statement is necessary (because as you point out it can only find the `log(int)` overload for that unqualified name and would thus try implicit conversions).

bbbales2 commented:
@seantalts Oh wow, thanks! Problem solved. I didn't notice that.

Yeah I was wondering how abs was working but ```using std::abs``` is right there too. I guess something must just be leaking the the mat.hpp header that is accidentally making this work for the vector case when it shouldn't.

bob-carpenter commented:
We don't need to do it for this pull request, but we really need to start cleaning up our includes and namespaces.  Developers keep trying to sneak in global using statements, which should be caught by cpplint but isn't.  That's where leaks come from.

When you have
```
  using std::log;
  log(3.2);  // std::log() available
  log(1);  // std::log() available
  std::log(3.2);  // std::log() forced
  std::log(1);  // std::log() forced
```
The using statement just the function `log()` (as defined in `std::log`) into namespace making it a candidate for resolution.  The most specific resolution will be used and it won't compile if there's not a unique most-specific match.

So wht happens when you have `std::log(x)` that doesn't happen when you have `using std::log; log(x)`?

bob-carpenter commented:
I should've added that if you're in `namespace stan { namespace math { ... } }` context then all of `stan::math` is automatically available.

bbbales2 commented:
> So wht happens when you have std::log(x) that doesn't happen when you have using std::log; log(x)?

Both of those should work. I'd messed up and removed the using std::log in one of my commits (this is what Sean pointed out). I switched to being explicit about std::log and std::abs.

On a related note, stan::math::log in prim doesn't seem to support doubles. stan::math::log in rev does (so the autodiff works). Is this considered a bug? If prim doesn't support what rev does?

bbbales2 commented:
This should be good to go.

seantalts commented:
(Please ignore the 3rd failing check - still working out the kinks)

bbbales2 commented:
Fixed the merge conflicts with the string stuff, this is good to go again.

bob-carpenter commented on stan/math/prim/scal/prob/cauchy_rng.hpp:
```diff
@@ -1,50 +1,60 @@
+      size_t N = max_size(mu, sigma);
+      VectorBuilder<true, double, T_loc, T_scale> output(N);
+
       check_finite(function, "Location parameter", mu);
```
Checks should go first before any work.  In general, try to put exceptional/odd/edge-case branches first so that when you're by them, you can scan the main branch of the program together.  Would you mind fixing this on all of these?  Sorry for not catchign that earlier.

bbbales2 commented:
Well you didn't see it before cause in the last commit when I was cleaning up whitespace n' such I just moved all the variable declarations to be in one blob :P. I've moved them below the argument checks now. Should be good to go if the tests pass!

bob-carpenter commented:
Thanks for moving up the argument validation tests.  I approved (don't know why I'm saying that since it's obvious in the interface).
